### ISSUE TITLE
Is the algorithm designed to improve with every training session?

### ISSUE THREAD

**Issue by ibraheemhlaiyil** [useaible/RyskampLearningMachine#16]
I've been testing the example apps and also implemented one of my own similar to the logistics app, in that I have no inputs, but need outputs that score best.

The algorithm tends towards a better set of outputs with lengthening network.NumSessions, however when the training session is completed, and the algorithm is run for a second training session, it appears as if it starts from scratch and does not "build on" the results of the previous training session? Is this correct behaviour?

I have set an initial randomness to less than 100% (I tried 90%) as I thought maybe initial randomness needs to be less than 100% in order to include results from the database from a previous training session, however the end score of the second training session is less than the first. Also, "predicting" at the last session of the training session does not use the "best solution" from the database from the previous training session, but uses the best solution of the current training session.

Perhaps I have misunderstood how the RLM works?

Thanks! And very interesting project! Hopefully it can be of use if I can understand it correctly!

**Comment by Jeff-Horton** [comment]
Hi Ibraheem,

Thanks for checking out the RLM, it's great to hear you're trying the technology.

The RLM is designed to continue learning from previous training runs, through the LoadNetwork functionality.

Are you calling RlmNetwork.LoadNetwork() after instantiating the RlmNetwork class but before configuring the object and running cycles.  Also, make sure that you only call RlmNetwork.NewNetwork if you are NOT continuing with a previous training run.

**Comment by randolph-useaible** [comment]
Hi Ibraheem,

Jeff is right as the RLM is able to continue learning from previous runs through the LoadNetwork and might not have been called. 

If that is not the case, then one thing to also make sure is if you are sending the correct database name for your RLM? You will see this as one of the parameters on the RlmNetwork constructor. Most of our examples, including the Logistics App, make use of a randomly generated database name so that it would start from scratch when running our examples. Simply change this to a static database name so that the RLM points to the same database every time and, in turn, loads the correct training data from previous runs.

Hope this solves your issue. Let us know if this helps.

**Comment by ibraheemhlaiyil** [comment]
Hi guys,
Thanks for the responses. I'm still working through this, it may well be that in my testing I overlooked ensuring that the database used was not not re-created with subsequent tests.
I have a further question.
It seems through the docs and test projects that the RLM is intended to be used serially, or synchronously, in the sense that one run is done before the score is used as feedback.
I would like to make use of all of the CPU cores I have available and run tests in parallel. What would be the best approach for this? Is it OK to run sessions in parallel?

**Comment by RixRyskamp** [comment]
Hi,

Good question.  Last year we changed our architectural model.  Instead of
using parallel sessions, each session runs in parallel under the covers and
should use all CPUs and NVIDIA GPUs.  You should see even processor usage
across your cores.  If you wish to use the GPU feature, please ask
@Randolph for support as this feature is not in the documentation yet.

If you feel like the RLM is not using all of your system resources
efficiently, please give us details.  It should.  We run our tests on a 40
core machine and it uses all cores.

Let me know how it works, I love the feedback.

Thanks,

Rix

CEO

**Comment by ibraheemhlaiyil** [comment]
Thanks @RixRyskamp 

I'll try to explain my problem and hopefully you can provide your advice.

I have various outputs generated at RunCycle and collected through CycleOutput.Outputs.

A session involves running RunCycle() once to produce the outputs. I have not checked how the CPU is utilised but this is not a concern for me.

These outputs are then run through a ScoreSession() function to generate a score for the session. The session is then ended with this score.

My problem is that ScoreSession is a very intensive, long-running function and is CPU-bound for the most part. As it is mainly synchronous I cannot run the ScoreSession function internally in parallel. This means that the session must wait for ScoreSession to finish before a new session can start, whilst other CPUs are idle.

Therefore I was wondering whether it would be OK to run the sessions themselves in parallel?

Thanks

Ibraheem

**Comment by RixRyskamp** [comment]
Thanks @ibraheemhlaiyil

I understand completely.  The problem is that the RLM learning is not
traditional and it requires the score to learn.  Starting a new session
before scoring the old one would not be productive because no learning
takes place without feedback.  Imagine playing a video game where you got
your score after 100 games.  It would be hard to learn to play that game.

My recommendation is that you work on parallelizing your scoring function.
My team has plenty of experience in making things parallel.  If you have
questions that are not too time-consuming I will authorize the team to help.

@Randolph, please Ibraheem with suggestions on making his scoring function
faster using parallel threads, etc.  Maybe point to places in the source
code that can help?

@ibraheemhlaiyil, what are you building??  If it is something we can
showcase later we can help more.

Thanks,

Rix

**Comment by ibraheemhlaiyil** [comment]
Thanks @RixRyskamp I thought this would be precisely the case but I thought I'd ask anyway in case there is a degree of randomness that might suggest outputs would be generated that might actually produce a better score (based on previous feedback + the randomness) than the one currently running. My thinking was that if this was the case, when the session was ended, this "better" feedback might then be used by the RLM internally. If the score was worse, no harm done, but if the score was better, we produced a better score faster than we would have if we ran the process without parallelisation.

Thanks for your help. I'll have to think how best to approach this problem, I might need to re-architect some of the code to do this. I'll let you know if I need any help though!

**Comment by RixRyskamp** [comment]
sounds good

**Comment by ibraheemhlaiyil** [comment]
Hi, revisiting this again after some time and still seeing this "issue".

I'm assuming that if network.TrainingDone() is called after 1000 session ends, the outputs found that produced the best score/solution, that those outputs would be used at the start of the next time the network is started up again for the next series of 1000 training sessions. However it doesn't look like this is the case? My start randomness is set to 100 and end randomness is 0. I have also tried start randomness of 90 and end of 0.

The best solution at the end of the second set of 1000 sessions, is not as good a score as the score of the best solution at the end of the first set of 1000 sessions.

I have checked and I am setting the network name the same with each session.

Another note, it seems like the EF has disk activity constantly after each session, is there a way to set disk activity to only occur at certain intervals? Like every X sessions, or every X seconds?

**Comment by randolph-useaible** [comment]
Hi,

I think what happened was that the underlying Data persistence process was not yet finished in saving the entire data hence the inaccurate best solution at the second training you did. It is normal for EF to have disk activity after each session because we are saving data in the background while the RLM is learning and for now, we do not have a way to delay it as you suggested. This was an effort to improve performance so that the RLM learning process would not have to wait for the DB persistence to finish saving each record. In turn, as the RLM learning process finishes faster, there is a tendency that the Data Persistence is lagging behind and might finish much later than the actual learning process depending on the volume of data being saved to the DB.

To know when the Data Persistence is done, you may subscribe to these events:

public event DataPersistenceCompleteDelegate DataPersistenceComplete;
public event DataPersistenceProgressDelegate DataPersistenceProgress;

Found on the [RlmNetwork](https://github.com/useaible/RyskampLearningMachine/blob/master/Core/RLM/RLM/RlmNetwork.cs) class.

With that, you will know when to correctly stop your application and ensure you have all the data persisted.

I hope this helps. 

Thanks

**Comment by ibraheemhlaiyil** [comment]
Thanks @randolph-useaible I will set that up and see if it makes a difference - also could you let me know what this function does exactly?

network.SetDataPersistenceProgressInterval();

Additionally, should I call network.TrainingDone() before or after the DataPersistenceComplete event, or does it not matter?

Thanks.

**Comment by ibraheemhlaiyil** [comment]
I've looked at the database and the following tables do not have any data saved at all:

Cases
Idea_Implementation
Idea_Module
Input_Values_Rneuron
Output_Values_Solution
RnetworkSettings
Rneurons
Solutions

These tables seem to have the correct data stored:

Input_Output_Type
Inputs
Outputs
Rnetwork
Sessions

So I can't see any data of actual "solutions" (outputs that generated the best score) being stored?

The scores are stored in the Sessions table correctly it seems.

**Comment by ibraheemhlaiyil** [comment]
I can confirm the DataPersistenceComplete event is called yet the tables that have no data remain empty.

**Comment by ibraheemhlaiyil** [comment]
I have worked out what the problem was - when passing inputs with values to `RunCycle`, I wasn't selecting an already-created input from the network object, but was creating a new input as when creating a new network.

However, I have noticed, not just with my application but with the demo applications also, that the `DataPersistenceComplete` event is not always fired, even if the application continues to run without a key press, so it is not always clear when it is safe to exit.

I have written a small loop after the call to `network.TrainingDone()` that checks for the completion of data persistence with upto a 10s wait:

            while (!dataPersistenceComplete)
            {
                Task.Delay(1000).Wait();

                if (DateTime.UtcNow.Subtract(LastDataPersistenceProgressUpdate).TotalSeconds > 10)
                {
                    dataPersistenceComplete = true;
                }
            }

The `DataPersistenceProgress` and `DataPersistenceComplete` event handlers keep a track of when the data was last updated and resets completion if it detected still pending data to be persisted.

        private void RlmNet_DataPersistenceProgress(long processed, long total)
        {
            LastDataPersistenceProgressUpdate = DateTime.UtcNow;

            if (total > processed)
            {
                dataPersistenceComplete = false;
            }

            if (showDataPersistProgress)
            {
                Console.WriteLine($"Data persistence progress - {processed} / {total}");
            }
        }

        private void RlmNet_DataPersistenceComplete()
        {
            dataPersistenceComplete = true;
            Console.WriteLine("Data persistence complete updated");
        }

I thought if there is a cleaner way of doing this you would have already done it!

Also - have you put any thought into using memory optimised tables as the EF data store?

**Comment by RixRyskamp** [comment]
Thank you for the detailed feedback.
